lib/data/plugins/markdown.js

```diff
@@ -1,27 +1,69 @@
-const Promise = require('bluebird')
-const Remark = Promise.promisifyAll(require('remark'))
-const html = require('remark-html')
-const highlight = require('remark-highlight.js')
-const remark = Remark()
+const _ = require('lodash');
+const Promise = require('bluebird');
+const remark = require('remark');
+
+const markdown = require('remark-parse');
+const html = require('remark-html');
+const slug = require('remark-slug');
+const highlight = require('remark-highlight.js');
+// remark-slug uses GithubSlugger, so we need it in order to verify
+// that our links will match the slugs generated
+const GithubSlugger = require('github-slugger');
+// We only care about `##` elements
+const SECTION_DEPTH = 2;
+
+const Remark = Promise.promisifyAll(remark);
+
+const formatPipe = Remark()
+  .use(markdown)
+  .use(slug)
   .use(highlight)
-  .use(html)
-const md = Promise.promisify(remark.process)
+  .use(html);
+
+const formatMarkdown = Promise.promisify(formatPipe.process);
 
 module.exports = {
   middleware: (store, action, next) => {
     if (action.payload.markdown) {
-      return md(action.payload.markdown).then(vfile => {
+      const slugger = new GithubSlugger();
+
+      const ast = remark.parse(action.payload.markdown);
+
+      const headings = _(ast.children)
+        .filter(x => x.type === 'heading' && x.depth === SECTION_DEPTH)
```

Comment:
I think this can be shortened to `.filter({ type: 'heading', depth: SECTION_DEPTH })`